sinonjs/sinon
# 🚀 Hello TypeScript

#### Purpose (TL;DR) - mandatory
Adds support for *gradually* migrating the sinon codebase over to TypeScript including the migration of a single utility and its test (`times-in-words`)

#### Background (Problem in detail)  - optional
Well seeing as it only took us the best part of 18 months to migrate to commonJS; I thought it would be worth a shot a migrating to TypeScript ;)  In all seriousness, I've found TypeScript to have provided a massive benefit to the projects that I've been working on making code written by other developers easier to reason about - in addition, using TypeScript allows us to make use of ES6+ features whilst still targeting ES5 browsers (heck, we can even target ES3!).  The other benefit is that once all source files are TypeScript (`.ts`, as opposed to `.js`), we can autogenerate the TypeScript definition files (`.d.ts`) which provides typehints to users of sinon - something will with be greatly appreciated by some in the community.

Obviously migrating the codebase to TypeScript is a pretty massive shift (and, ultimately to some degree, boils down to a personal preference) - hence why I've done an absolute MVP here in this case: however I'm keen to explore what the 'guts' of sinon will look like typed.

Finally, for those new to the TypeScript eco-system the ES5 code generated by the TypeScript compiler is very clean (and can easily be mistaken for hand-written ES5 once formatted) - so a migration back to plan javascript in the future would be reasonably trivial (ie: minimal sunk cost).

#### Solution  - optional
- Add typescript and tslint; use a sensible `tsconfig.json` which allows interop with the existing javascript source
- Move all sinon source code in `src/` and transpile it into the `lib` folder at build time.
- Run `tsc` before each test suite - this is incredibly inefficient, however compilation speed is very fast at this stage and I didn't want to mess around too much for the MVP: this can easily be optimised later if we wish to invest in this approach.

#### How to verify - mandatory
1. Check out this branch (see github instructions below)
2. `npm install`
3. npm test
